### ISSUE TITLE
Creating document/text Vectors

### ISSUE THREAD

**Issue by SyedShahzaib790** [oduwsdl/sumgram#23]
Hi there,
First of all, I am really glad that I found this tool. I am pretty sure that we can use this tool to create vectors for a given text/document. My question is: In order to create vectors using sumgram, do we have to pass each document one by one? and then have to apply the same technique we normally do for a "bag of words" approach or is there any better way of converting a bunch of documents to their respective numerical vectors using this tool. Also, can we apply TF-IDF on those vectors?

**Comment by anwala** [comment]
Hello @SyedShahzaib790 

sumgram uses a vector with TF/DF weights as primitives to generate sumgrams. But if you just want to generate vectors from documents consisting of TFs or TF-IDF weights, sumgram is the wrong tool. Instead you can achieve this more directly by using the library sumgram uses [sklearn's countvectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) to generate vectors of TF or TF-IDF weights for documents.

Here are some simple examples (TF matrix (un)normalized, TF-IDF, etc.) that assume all your text documents are in the same list (doc_list) even though countvectorizer seems to be able to take iterators as inputs.
```
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.preprocessing import normalize
import numpy as np

stop_words = {'the', 'a', 'them', 'with', 'is', 'of', 'over', 'has'}
token_pattern = r'(?u)\b[a-zA-Z\'\’-]+[a-zA-Z]+\b|\d+[.,]?\d*'
ngram_range = (1, 1)

doc_list = [
    'The eye of Category 4 Hurricane Harvey is now over Aransas Bay.',
    'Eye of Category 4 Hurricane Harvey is almost onshore.',
    'Hurricane Harvey has become a Category 4 storm with maximum sustained winds of 130 mph.'
]

count_vectorizer = CountVectorizer(token_pattern=token_pattern, stop_words=stop_words, ngram_range=ngram_range)
tf_mat = count_vectorizer.fit_transform(doc_list).toarray()

print('Vocabulary:')
print( count_vectorizer.get_feature_names() )
print()

print('Raw TF (not normalized) matrix')
print( tf_mat )
print()

tf_mat_normalized = normalize(tf_mat, norm='l2', axis=1)
print('TF (normalized with l2 norm) matrix')
print( tf_mat_normalized )
print()

tfidf = TfidfTransformer(norm='l2')
tfidf.fit(tf_mat)
tf_idf_matrix = tfidf.transform(tf_mat).todense()
print('TF-IDF (normalized with l2 norm) matrix')
print(tf_idf_matrix)
print()

def most_freq_ngrams(count_vectorizer, tf_mat):

    top_freq_ngrams = []
    vocab = count_vectorizer.get_feature_names()
    all_col_sums_tf = np.sum(tf_mat, axis=0)
    
    for i in range( len(vocab) ):
        top_freq_ngrams.append( {'term': vocab[i], 'tf': int(all_col_sums_tf[i])} )
    
    top_freq_ngrams = sorted(top_freq_ngrams, key=lambda x: x['tf'], reverse=True)
    print('MOST frequent ngrams')
    for ngram in top_freq_ngrams:
        print(ngram)

most_freq_ngrams(count_vectorizer, tf_mat)
```

**Comment by SyedShahzaib790** [comment]
Hi @anwala 
Thanks for the quick and detailed response. I totally get the idea of using TF-IDF alone but I wanted to somehow combine your idea of sumgram to build a vocabulary of more meaningful phrases instead of going through the ngram phrases sequentially. You see, in the above code, we are treating every word (or list of words) equally and generating our vocabulary. What if, our vocabulary is provided by the output of sumgram to countVectorizer.... Are you getting my point? or am I missing something?

Thanks

**Comment by anwala** [comment]
Hello again @SyedShahzaib790 

I'm happy to help if possible. sumgram generates conjoined ngrams, e.g.,

1. `centers for disease control and prevention`: 6-gram
2. `public health`: 2-gram
3. `the world health organization`: 4-gram

Technically, you could collect the top k (e.g, k=100) sumgrams generated by sumgram and send that to CountVectorizer. I can think of 2 ways to do this. The first way doesn't make a lot of sense to me. The second makes more sense to me:

1. You may have to combine all the sumgrams into a single document and pass it to CountVectorizer. But that would split the conjoined ngrams (sumgrams) invalidating the hard work sumgrams did to join the ngrams in the first place. But depending on your context, this is possible. This is illustrated by the code from my first comment.

2. You can pass CountVectorizer your own [`vocabulary`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) (see vocabulary parameter). This means CountVectorizer would not determine the vocabulary from `input`. Your vocabulary from sumgrams would consist of multiple ngrams (2-gram, 3-gram, 4-gram, etc). 

Method 2 above seems most suitable for your task, to the best of my understanding.
Good luck.

**Comment by SyedShahzaib790** [comment]
Hi @anwala,
Exactly, that what I was expecting from you in terms of confirmation. Absolutely, the method looks more convincing and  I too had a similar idea in mind. Now, in order to get K sumgrams I would be passing the whole text at a time (list of documents) in a single call, right? correct me if I am wrong and later I will eventually create vectors of size K using those vocabulary generated by sumgram.

For a test document/query, I will get the sumgrams and would create a vector of similar size (i.e. K). Now this can be used for text similarity usecase too. Intuitively, I think it should work better than a simple bag of words approach for text similarity usecase.

**Comment by anwala** [comment]
Yes; currently sumgram needs all the text documents to be passed in a single list. We plan to implement a more efficient method that would not require all the documents to reside in memory. But for now, all the documents have to be passed in a single list. 

See the example in the documentation for how to pass a list consisting of text documents to extract sumgrams: [python-script-usage](https://github.com/oduwsdl/sumgram#python-script-usage). In the example, k = `top_sumgram_count` = 10. But since you want to use k sumgrams as a vocabulary, I suppose k has to be bigger e.g., (k = 100 or k=1000).

Next, you'd extract the k sumgrams and create a vocabulary and send to CountVectorizer (along with the document to process).

Two documents `A` and `B` can only be compared using vector-based or set-based similarity applied to their respective sumgrams. I prefer (and have implemented) the set-based method.

**Method 1 (Vector-based similarity): Doc comparison using sumgrams**
If you independently process `A` and `B`. You'd get two possibly different vocabulary even though they might be from the same topic. 
E.g., `A`'s vocabulary could look like this:
1. `ebola virus`
2. `centers for disease control and prevention`
3. `outbreak in west africa`
4. `public health crisis`

`B`'s vocabulary could look like this:
1. `ebola outbreak in west africa`
2. `guinea liberia sierra`
3. `public health crisis`
4. `ebola virus`

Given that both documents have two different vocabularies (extracted from sumgram) they cannot be compared using vector-based method (cosine similarity) until you combine (e.g., union) their respective vocabulary (represent both in the same vector-space). But I don't think this is necessary, you can just compute the similarity using a set-based method.

**Method 2 (Set-based similarity): Doc comparison using sumgrams**
With set-based method, all you have to do is compute a set similarity (either Jaccard, Overlap coeff., Jaccard+Overlap, etc) between the sets of `A` and `B`'s sumgrams. In fact I have used this method before to compare documents by just using their sumgrams. See sample code below that uses overlap as a primitive set-based doc comparison method:
```
def overlapFor2Sets(firstSet, secondSet):

    intersection = float(len(firstSet & secondSet))
    minimum = min(len(firstSet), len(secondSet))

    if( minimum != 0 ):
        return  round(intersection/minimum, 4)
    else:
        return 0

def customCountIntsct(aSetLst, bSetLst):

    commonCount = 0

    for aSet in aSetLst:
        for bSet in bSetLst:
            if( overlapFor2Sets(aSet, bSet) == 1 ):
                #print('\tMATCH', aSet, '<===>' , bSet)
                commonCount += 1

    return commonCount

def sumgramSim(aCol, bCol, k=10, verbose=False):
    
    minLen = min(len(aCol), len(bCol))
    if( k > minLen ):
        k = minLen

    if( k == 0 ):
        return 0


    aCol = aCol[:k]
    bCol = bCol[:k]

    aSetLst = []
    bSetLst = []
    
    if( verbose is True ):
        print('k:', k)

    for i in range(k):

        if( verbose ):
            print(aCol[i], '<===>', bCol[i])
        
        aSetLst.append( set(aCol[i].split(' ')) )
        bSetLst.append( set(bCol[i].split(' ')) )

    if( verbose ):
        print()
        for i in range(k):
            print(aSetLst[i], '<===>', bSetLst[i])      

    commonCount = customCountIntsct(aSetLst, bSetLst)
    sim = commonCount/k
    
    if( verbose ):
        print('sim:', sim)
        print('\n\n')

    return sim

fst = [
    'ebola virus',
    'public health crisis'
]
sec = [
    'ebola virus',
    'public health crisis'
]
print('sumgram sim for same text:', sumgramSim(fst, sec))

fst = [
    'ebola virus',
    'centers for disease control and prevention',
    'outbreak in west africa',
    'public health crisis'
]
sec = [
    'ebola outbreak in west africa',
    'guinea liberia sierra',
    'public health crisis',
    'ebola virus'
]
print('sumgram sim for text:', sumgramSim(fst, sec))
```

**Comment by SyedShahzaib790** [comment]
@anwala 
Great! Thank you very much for the answer and code above. I am definitely going to try these methods and will see which works better.

Thanks again, Stay safe!
